src/turbomind/models/llama/LlamaBatch.cc

```diff
@@ -260,21 +260,27 @@ void LlamaBatch<T>::ProcessInferRequests(const Requests& requests)
             }
         }
 
+        // update stage and interact count for stateful inference
+        seq.stage = Sequence::kPrefill;
+        seq.interact_count++;
+
+        const int history_length = seq.tokens.size();
+
         const int  input_length = r->inputs[rank_].getVal<int>("input_lengths");
         const int* input_ids    = r->inputs[rank_].getPtr<int>("input_ids");
 
         // `output_ids` contains all token ids of the sequences
         const auto output_ids_base = state.output_ids + session_len_ * idx;
         auto       output_ids      = output_ids_base;
 
-        // copy history tokens
-        if (!seq.tokens.empty()) {
-            output_ids = Copy(seq.tokens.data(), seq.tokens.size(), output_ids);
+        if (input_length) {
+            // update tokens in sequence
+            seq.tokens.resize(history_length+input_length);
+            std::copy_n(input_ids, input_length, &seq.tokens[history_length]);
```

Comment:
Could you point out what sequence state would be corrupted? And any suggestions to handle this?

Reply:
> Could you point out what sequence state would be corrupted?

`seq.tokens` will be inconsistent with what's filled in the KV cache.

> And any suggestions to handle this?

May add a `prompt` field for `Sequence` to hold a copy of input ids for newly created sequence (`start_flag == true `). It can be cleared after `CacheIfEnabled`. This way both `stage` and `interact_count` can be eliminated.

Also, be careful when there exists `input_embeddings`. Prefix matching must avoid matching embedding tokens as the supplied token ids are duplicated model specific dummy id. So the copying of prompt tokens stops at first embedding token.

@irexyc What's the best way to get the first occurance of embedding tokens in `ProcessInferRequests`?

Reply:
> @irexyc What's the best way to get the first occurance of embedding tokens in `ProcessInferRequests`?

We could only enable Prefix Cache for LLM and not use for VLM.

Reply:
> updating seq.tokens here will currupt the sequence state if it gets canceled before finishing. 

Do you mean stop requests? If the sequence gets canceled before finishing, the sequence will be interrupted and erased.

Reply:
Stop request only stop the current request, the sequence won't be earsed. It can be resumed from any step in it's history.